src/eduid_webapp/security/views/webauthn.py

```diff
@@ -0,0 +1,165 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import print_function, absolute_import, unicode_literals
+
+import struct
+import json
+import base64
+from flask import Blueprint, session, Response
+from flask import current_app, request
+
+from fido2.client import ClientData
+from fido2.server import Fido2Server, RelyingParty
+from fido2.ctap2 import AttestationObject
+from fido2 import cbor
+from fido2.utils import websafe_encode
+from fido2.ctap2 import AttestedCredentialData
+
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from OpenSSL import crypto
+
+from eduid_userdb.credentials import Webauthn
+from eduid_userdb.security import SecurityUser
+from eduid_common.api.decorators import require_user, MarshalWith, UnmarshalWith
+from eduid_common.api.utils import save_and_sync_user
+from eduid_webapp.security.helpers import credentials_to_registered_keys, compile_credential_list
+from eduid_webapp.security.schemas import WebauthnOptionsResponseSchema, WebauthnRegisterRequestSchema
+from eduid_webapp.security.schemas import SecurityResponseSchema, RemoveWebauthnTokenRequestSchema
+from eduid_webapp.security.schemas import VerifyWithWebauthnTokenRequestSchema
+from eduid_webapp.security.schemas import VerifyWithWebauthnTokenResponseSchema
+
+
+WEBAUTHN_SERVER = None
+
+def update_webauthn_server(rp_id, name='eduID security API'):
+    rp = RelyingParty(rp_id, name)
+    server = Fido2Server(rp)
+    global WEBAUTHN_SERVER
+    WEBAUTHN_SERVER = server
+    return server
+
+def get_webauthn_server():
+    if WEBAUTHN_SERVER is not None:
+        return WEBAUTHN_SERVER
+    return update_webauthn_server(current_app.config['FIDO2_RP_ID'])
+
+
+def make_credentials(creds):
+    credentials = []
+    for cred in creds:
+        cred_data = base64.b64decode(cred.credential_data.encode('ascii'))
+        credential_data, rest = AttestedCredentialData.unpack_from(cred_data)
+        if rest:
+            continue
+        credentials.append(credential_data)
+    return credentials
+
+
+webauthn_views = Blueprint('webauthn', __name__, url_prefix='/webauthn', template_folder='templates')
+
+@webauthn_views.route('/register/begin', methods=['GET'])
+@MarshalWith(WebauthnOptionsResponseSchema)
```

Comment:
MarshalWith will be short-circuited by returning a Response. I wonder if we should teach MarshalWith about cbor or skip using the MarshalWith decorator. Maybe it could still be used as documentation,

Reply:
removed